core/services/ocr3/plugins/ccip/internal/reader/onchain_prices_reader.go

```diff
@@ -0,0 +1,66 @@
+package reader
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	ocr2types "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	"golang.org/x/sync/errgroup"
+)
+
+type TokenPriceConfig struct {
+	// This is mainly used for inputTokens on testnet to give them a price
+	StaticPrices map[ocr2types.Account]big.Int `json:"staticPrices"`
+}
+
+type OnchainTokenPricesReader struct {
+	TokenPriceConfig TokenPriceConfig
+	// Reader for the chain that will have the token prices on-chain
+	ContractReader commontypes.ContractReader
+}
+
+func NewOnchainTokenPricesReader(tokenPriceConfig TokenPriceConfig, contractReader commontypes.ContractReader) *OnchainTokenPricesReader {
+	return &OnchainTokenPricesReader{
+		TokenPriceConfig: tokenPriceConfig,
+		ContractReader:   contractReader,
+	}
+}
+
+func (pr *OnchainTokenPricesReader) GetTokenPricesUSD(ctx context.Context, tokens []ocr2types.Account) ([]*big.Int, error) {
+	const (
+		contractName = "PriceAggregator"
+		functionName = "getTokenPrice"
+	)
+	prices := make([]*big.Int, len(tokens))
+	eg := new(errgroup.Group)
+	for idx, token := range tokens {
+		idx := idx
+		token := token
+		eg.Go(func() error {
+			price := new(big.Int)
+			if staticPrice, exists := pr.TokenPriceConfig.StaticPrices[token]; exists {
+				price.Set(&staticPrice)
+			} else {
+				if err := pr.ContractReader.GetLatestValue(ctx, contractName, functionName, token, price); err != nil {
+					return fmt.Errorf("failed to get token price for %s: %w", token, err)
+				}
+			}
+			prices[idx] = price
+			return nil
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, fmt.Errorf("failed to get all token prices successfully: %w", err)
+	}
+
+	for _, price := range prices {
+		if price == nil {
+			return nil, fmt.Errorf("failed to get all token prices successfully, some prices are nil")
+		}
+	}
```

Comment:
What do you think about including the nil tokens in this error message? i.e.
```suggestion
	var nilTokens []ocr2types.Account
	for i, price := range prices {
		if price == nil {
			nilTokens = append(nilTokens, tokens[i])
		}
	}
	if len(nilTokens) != 0 {
		return nil, fmt.Errorf("failed to get all token prices successfully, some prices are nil: %s", strings.Join(nilTokens, ","))
	}
```

Reply:
oops, just merged 😅. I like the idea. I'll make another simple PR for it once I finish the home chain config comments. 👍